File: results/dnd-rest-benchmark/lifecycle-runs/20260804T072004Z_claude_claude-sonnet-5_ruby-sinatra/routes/npcs.rb
# Campaign NPCs, factions, and a derived relationship summary.

STANCES = %w[friendly neutral hostile].freeze

post '/v1/campaigns/:campaign_id/factions' do
  halt 404, { error: 'unknown campaign' }.to_json unless campaign_exists?(params[:campaign_id])

  body = json_body
  halt 400, { error: 'invalid json' }.to_json if body.nil?

  id = body['id']
  name = body['name']
  stance = body['stance']

  halt 400, { error: 'invalid id' }.to_json unless id.is_a?(String) && !id.empty?
  halt 400, { error: 'invalid name' }.to_json unless name.is_a?(String) && !name.empty?
  halt 400, { error: 'invalid stance' }.to_json unless STANCES.include?(stance)
  if db.execute('SELECT 1 FROM campaign_factions WHERE campaign_id = ? AND id = ?', [params[:campaign_id], id]).first
    halt 409, { error: 'id already exists' }.to_json
  end

  db.execute(
    'INSERT INTO campaign_factions (campaign_id, id, name, stance) VALUES (?, ?, ?, ?)',
    [params[:campaign_id], id, name, stance]
  )

  status 201
  { id: id, name: name, stance: stance }.to_json
end

post '/v1/campaigns/:campaign_id/npcs' do
  halt 404, { error: 'unknown campaign' }.to_json unless campaign_exists?(params[:campaign_id])

  body = json_body
  halt 400, { error: 'invalid json' }.to_json if body.nil?

  id = body['id']
  name = body['name']
  faction_id = body['faction_id']
  disposition = body['disposition']

  halt 400, { error: 'invalid id' }.to_json unless id.is_a?(String) && !id.empty?
  halt 400, { error: 'invalid name' }.to_json unless name.is_a?(String) && !name.empty?
  unless faction_id.nil? || (faction_id.is_a?(String) && !faction_id.empty?)
    halt 400, { error: 'invalid faction_id' }.to_json
  end
  halt 400, { error: 'invalid disposition' }.to_json unless integerish(disposition)
  if !faction_id.nil? && !db.execute('SELECT 1 FROM campaign_factions WHERE campaign_id = ? AND id = ?', [params[:campaign_id], faction_id]).first
    halt 404, { error: 'unknown faction' }.to_json
  end
  if db.execute('SELECT 1 FROM campaign_npcs WHERE campaign_id = ? AND id = ?', [params[:campaign_id], id]).first
    halt 409, { error: 'id already exists' }.to_json
  end

  disposition = disposition.to_i

  db.execute(
    'INSERT INTO campaign_npcs (campaign_id, id, name, faction_id, disposition) VALUES (?, ?, ?, ?, ?)',
    [params[:campaign_id], id, name, faction_id, disposition]
  )

  status 201
  { id: id, name: name, faction_id: faction_id, disposition: disposition }.to_json
end

get '/v1/campaigns/:campaign_id/relationships' do
  halt 404, { error: 'unknown campaign' }.to_json unless campaign_exists?(params[:campaign_id])

  factions = db.execute('SELECT COUNT(*) AS cnt FROM campaign_factions WHERE campaign_id = ?', [params[:campaign_id]]).first['cnt']
  npcs = db.execute('SELECT COUNT(*) AS cnt FROM campaign_npcs WHERE campaign_id = ?', [params[:campaign_id]]).first['cnt']
  friendly_npcs = db.execute(
    'SELECT COUNT(*) AS cnt FROM campaign_npcs WHERE campaign_id = ? AND disposition > 0',
    [params[:campaign_id]]
  ).first['cnt']

  {
    campaign_id: params[:campaign_id],
    factions: factions,
    npcs: npcs,
    friendly_npcs: friendly_npcs
  }.to_json
end

File: results/dnd-rest-benchmark/lifecycle-runs/20260731T062307Z_claude_claude-sonnet-5_typescript-node/src/routes/npcs.ts
// Campaign NPCs and factions, plus a relationship summary derived from
// their disposition state. Persistent (`campaign_factions`, `campaign_npcs`).
import type { ServerResponse } from "node:http";
import { db } from "../db.js";
import { sendJson } from "../http.js";
import { isPlainObject, isValidInt } from "../validation.js";
import { hasCampaign } from "./campaigns.js";

const VALID_STANCES = new Set(["friendly", "neutral", "hostile"]);

interface FactionRecord {
  campaignId: string;
  id: string;
  name: string;
  stance: string;
}

interface NpcRecord {
  campaignId: string;
  id: string;
  name: string;
  factionId: string | null;
  disposition: number;
}

function hasFaction(campaignId: string, id: string): boolean {
  const row = db.prepare("SELECT 1 FROM campaign_factions WHERE campaign_id = ? AND id = ?").get(campaignId, id);
  return row !== undefined;
}

function saveFaction(faction: FactionRecord): void {
  db.prepare("INSERT INTO campaign_factions (campaign_id, id, name, stance) VALUES (?, ?, ?, ?)").run(
    faction.campaignId,
    faction.id,
    faction.name,
    faction.stance,
  );
}

function hasNpc(campaignId: string, id: string): boolean {
  const row = db.prepare("SELECT 1 FROM campaign_npcs WHERE campaign_id = ? AND id = ?").get(campaignId, id);
  return row !== undefined;
}

function saveNpc(npc: NpcRecord): void {
  db.prepare(
    "INSERT INTO campaign_npcs (campaign_id, id, name, faction_id, disposition) VALUES (?, ?, ?, ?, ?)",
  ).run(npc.campaignId, npc.id, npc.name, npc.factionId, npc.disposition);
}

function countFactions(campaignId: string): number {
  const row = db.prepare("SELECT COUNT(*) AS count FROM campaign_factions WHERE campaign_id = ?").get(
    campaignId,
  ) as { count: number };
  return row.count;
}

function getNpcs(campaignId: string): NpcRecord[] {
  const rows = db
    .prepare("SELECT campaign_id, id, name, faction_id, disposition FROM campaign_npcs WHERE campaign_id = ?")
    .all(campaignId) as { campaign_id: string; id: string; name: string; faction_id: string | null; disposition: number }[];
  return rows.map((row) => ({
    campaignId: row.campaign_id,
    id: row.id,
    name: row.name,
    factionId: row.faction_id,
    disposition: row.disposition,
  }));
}

export function handleCreateFaction(res: ServerResponse, campaignId: string, body: unknown): void {
  if (!hasCampaign(campaignId)) {
    sendJson(res, 404, { error: "campaign not found" });
    return;
  }

  if (
    !isPlainObject(body) ||
    typeof body.id !== "string" ||
    !body.id ||
    typeof body.name !== "string" ||
    typeof body.stance !== "string" ||
    !VALID_STANCES.has(body.stance)
  ) {
    sendJson(res, 400, { error: "invalid request" });
    return;
  }

  if (hasFaction(campaignId, body.id)) {
    sendJson(res, 409, { error: "faction already exists" });
    return;
  }

  const faction: FactionRecord = { campaignId, id: body.id, name: body.name, stance: body.stance };
  saveFaction(faction);

  sendJson(res, 201, { id: faction.id, name: faction.name, stance: faction.stance });
}

export function handleCreateNpc(res: ServerResponse, campaignId: string, body: unknown): void {
  if (!hasCampaign(campaignId)) {
    sendJson(res, 404, { error: "campaign not found" });
    return;
  }

  if (
    !isPlainObject(body) ||
    typeof body.id !== "string" ||
    !body.id ||
    typeof body.name !== "string" ||
    typeof body.faction_id !== "string" ||
    !body.faction_id ||
    !isValidInt(body.disposition, -5, 5)
  ) {
    sendJson(res, 400, { error: "invalid request" });
    return;
  }

  if (!hasFaction(campaignId, body.faction_id)) {
    sendJson(res, 404, { error: "faction not found" });
    return;
  }

  if (hasNpc(campaignId, body.id)) {
    sendJson(res, 409, { error: "npc already exists" });
    return;
  }

  const npc: NpcRecord = {
    campaignId,
    id: body.id,
    name: body.name,
    factionId: body.faction_id,
    disposition: body.disposition,
  };
  saveNpc(npc);

  sendJson(res, 201, {
    id: npc.id,
    name: npc.name,
    faction_id: npc.factionId,
    disposition: npc.disposition,
  });
}

export function handleRelationshipSummary(res: ServerResponse, campaignId: string): void {
  if (!hasCampaign(campaignId)) {
    sendJson(res, 404, { error: "campaign not found" });
    return;
  }

  const npcs = getNpcs(campaignId);
  const friendlyNpcs = npcs.filter((npc) => npc.disposition > 0).length;

  sendJson(res, 200, {
    campaign_id: campaignId,
    factions: countFactions(campaignId),
    npcs: npcs.length,
    friendly_npcs: friendlyNpcs,
  });
}
